02-peer-protocol.md

```diff
@@ -1713,6 +1784,181 @@ satoshis, which is possible if `dust_limit_satoshis` is below 546 satoshis).
+   * [`closer_scriptpubkey_len*byte`:`closer_scriptpubkey`]
+   * [`u16`:`closee_scriptpubkey_len`]
+   * [`closee_scriptpubkey_len*byte`:`closee_scriptpubkey`]
+   * [`u64`:`fee_satoshis`]
+   * [`u32`:`locktime`]
+   * [`closing_tlvs`:`tlvs`]
+
+1. type: 41 (`closing_sig`)
+2. data:
+   * [`channel_id`:`channel_id`]
+   * [`u16`:`closer_scriptpubkey_len`]
+   * [`closer_scriptpubkey_len*byte`:`closer_scriptpubkey`]
+   * [`u16`:`closee_scriptpubkey_len`]
+   * [`closee_scriptpubkey_len*byte`:`closee_scriptpubkey`]
+   * [`u64`:`fee_satoshis`]
+   * [`u32`:`locktime`]
+   * [`closing_tlvs`:`tlvs`]
+
+1. `tlv_stream`: `closing_tlvs`
+2. types:
+    1. type: 1 (`closer_output_only`)
+    2. data:
+        * [`signature`:`sig`]
+    1. type: 2 (`closee_output_only`)
+    2. data:
+        * [`signature`:`sig`]
+    1. type: 3 (`closer_and_closee_outputs`)
+    2. data:
+        * [`signature`:`sig`]
+
+#### Requirements
+
+Note: the details and requirements for the transaction being signed are in [BOLT 3](03-transactions.md#closing-transaction).
+
+An output is *dust* if the amount is less than the [Bitcoin Core Dust Thresholds](03-transactions.md#dust-limits).
+
+Both nodes:
+  - After a `shutdown` has been sent and received, AND no HTLCs remain in either commitment transaction:
+    - SHOULD send a `closing_complete` message.
+
+The sender of `closing_complete` (aka. "the closer"):
+  - MUST set `fee_satoshis` to a fee less than or equal to its outstanding balance, rounded down to whole satoshis.
+  - MUST set `fee_satoshis` so that at least one output is not dust.
+  - MUST set `closer_scriptpubkey` to its desired output script.
+  - MUST set `closee_scriptpubkey` to the last script it received from its peer (from `closing_complete` or from the initial `shutdown`).
+  - MUST set `locktime` to the desired `nLockTime` of the closing transaction.
+  - MUST generate its closing transaction as specified in [BOLT #3](03-transactions.md#closing-transaction).
+  - MUST set `signature` fields as valid signature using its `funding_pubkey` of:
+    - `closer_output_only`: closing transaction with only the local ("closer") output.
+    - `closee_output_only`: closing transaction with only the remote ("closee") output.
+    - `closer_and_closee_outputs`: closing transaction with both the closer and closee outputs.
+  - If the local outstanding balance (in millisatoshi) is less than the remote outstanding balance:
+    - MUST NOT set `closer_output_only`.
+    - MUST set `closee_output_only` if the local output amount is dust.
+    - MAY set `closee_output_only` if it considers the local output amount uneconomical AND its `closer_scriptpubkey` is not `OP_RETURN`.
+  - Otherwise (not lesser amount, cannot remove its own output):
+    - MUST NOT set `closee_output_only`.
+    - If the closee's output amount is dust:
+      - MUST set `closer_output_only`.
+      - SHOULD NOT set `closer_and_closee_outputs`.
+    - Otherwise:
```

Comment:
Right now the `OP_RETURN` logic is only detailed at the end in the rationale section. I think we should add a requirement here to make it clearer _when_ someone can/should send a script with an `OP_RETURN`. So maybe something like:
```
- If the closer considers their output uneconomical, and the output of the closee is dust:
  - The closer MAY send a `closer_scriptpubkey` that is a valid `OP_RETURN` script. 
    - If sent, then the closer MUST set their output value to zero, ensuring that only a single output exists on the closing transaction (all funds go to fees). 

```